src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs

```diff
@@ -116,16 +116,19 @@ public static SocketError Connect(SafeCloseSocket handle, byte[] peerAddress, in
                 IntPtr.Zero);
             return errorCode == SocketError.SocketError ? GetLastSocketError() : SocketError.Success;
         }
-
         public static SocketError Send(SafeCloseSocket handle, IList<ArraySegment<byte>> buffers, SocketFlags socketFlags, out int bytesTransferred)
         {
+            const int StackThreshold = 80; // arbitrary limit to avoid too much space on stack
             int count = buffers.Count;
-            WSABuffer[] WSABuffers = new WSABuffer[count];
-            GCHandle[] objectsToPin = null;
+            Span<WSABuffer> WSABuffers = count < StackThreshold ? stackalloc WSABuffer[count] : new WSABuffer[count];
+            Span<GCHandle> objectsToPin = count < StackThreshold ? stackalloc GCHandle[count] : new GCHandle[count];
```

Comment:
Did you consider using `ArrayPool` for the fallback path instead of allocating a new?